internal/wasm/wazero_sdk/internal/wazero/VM.go

```diff
@@ -0,0 +1,516 @@
+//Is used when setting the policy data
+//
+func (i *VM) toDRegoJSON(ctx context.Context, v interface{}, free bool) error {
+	var raw []byte
+	switch v := v.(type) {
+	case []byte:
+		raw = v
+	case *ast.Term:
+		raw = []byte(v.String())
+	case ast.Value:
+		raw = []byte(v.String())
+	default:
+		var err error
+		raw, err = json.Marshal(v)
+		if err != nil {
+			return err
+		}
+	}
+
+	n := int32(len(raw))
+	p := int32(i.module.writeMem(raw))
+	i.dataLen = int32(n)
+	addr, err := i.valueParse(ctx, p, n)
+	if err != nil {
+		return err
+	}
+	i.dataAddr = addr
+	cPtr, err := i.getHeapState(ctx)
+	if err != nil {
+		return err
+	}
+	i.dataLen = cPtr - addr
+	if free {
+		if err := i.free(ctx, p); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+func (i *VM) getHeapState(ctx context.Context) (int32, error) {
+	return i.heapPtrGet(ctx)
+}
+
+func (i *VM) setHeapState(ctx context.Context, ptr int32) error {
+	return i.heapPtrSet(ctx, ptr)
+}
+
+//copies the parsed data to optimize cloning VMs
+func (i *VM) cloneDataSegment() (int32, []byte) {
+	srcData := i.module.readFrom(0)[i.baseHeapPtr:i.evalHeapPtr]
+	patchedData := make([]byte, len(srcData))
+	copy(patchedData, srcData)
+	return i.dataAddr, patchedData
+}
+func (i *VM) GetEntrypoints() map[string]int32 {
+	return i.module.GetEntrypoints()
+}
+func (i *VM) Eval(ctx context.Context,
+	entrypoint int32,
+	input *interface{},
+	metrics metrics.Metrics,
+	seed io.Reader,
+	ns time.Time,
+	iqbCache cache.InterQueryCache,
+	ph print.Hook,
+	capabilities *ast.Capabilities) ([]byte, error) {
+	i.module.Reset(ctx, seed, ns, iqbCache, ph, capabilities)
+	if i.abiMinorVersion < int32(2) {
+		return i.evalCompat(ctx, entrypoint, input, metrics, seed, ns, iqbCache, ph, capabilities)
+	}
+
+	metrics.Timer("wasm_vm_eval").Start()
+	defer metrics.Timer("wasm_vm_eval").Stop()
+
+	inputAddr, inputLen := int32(0), int32(0)
+
+	// NOTE: we'll never free the memory used for the input string during
+	// the one evaluation, but we'll overwrite it on the next evaluation.
+	heapPtr := i.evalHeapPtr
+
+	if input != nil {
+		metrics.Timer("wasm_vm_eval_prepare_input").Start()
+		var raw []byte
+		switch v := (*input).(type) {
+		case []byte:
+			raw = v
+		case *ast.Term:
+			raw = []byte(v.String())
+		case ast.Value:
+			raw = []byte(v.String())
+		default:
+			var err error
+			raw, err = json.Marshal(v)
+			if err != nil {
+				return nil, err
+			}
+		}
+		inputLen = int32(len(raw))
+		inputAddr = i.evalHeapPtr
+		err := i.module.writeMemPlus(uint32(inputAddr), raw, "input")
+		if err != nil {
+			return nil, err
+		}
+		heapPtr += inputLen
+		metrics.Timer("wasm_vm_eval_prepare_input").Stop()
+	}
+	// Setting the ctx here ensures that it'll be available to builtins that
+	// make use of it (e.g. `http.send`); and it will spawn a go routine
+	// cancelling the builtins that use topdown.Cancel, when the context is
+	// cancelled.
+	i.module.Reset(ctx, seed, ns, iqbCache, ph, capabilities)
```

Comment:
I'm afraid it's been a while since I've worked on the wasmtime-go integration -- so please bear with me -- Is it the right thing to do to reset the rng seed on every Eval?

Reply:
It is what the wasmtime sdk does, although I have to assume that seed is passed as nil and is thus set to ns in unix timestamp format